simulator/func_sim/alu.h

```diff
@@ -10,12 +10,84 @@
 #include "multiplication.h"
 #include "traps/trap.h"
 
+#include <array>
 #include <infra/macro.h>
 #include <tuple>
 
 template<size_t N, typename T>
 T align_up(T value) { return ((value + ((1ull << N) - 1)) >> N) << N; }
 
+struct Mask {
+    const uint32 left_mask, right_mask;
+};
+
+static std::array<Mask, 4> shuffle_consts = {
+    Mask{ 0x00ff0000, 0x0000ff00},
+    Mask{ 0x0f000f00, 0x00f000f0},
+    Mask{ 0x30303030, 0x0c0c0c0c},
+    Mask{ 0x44444444, 0x22222222},
+};
+
+static uint32 shuffle_stage( uint32 src, size_t index)
+{
+    uint32 maskL = shuffle_consts[index].left_mask;
+    uint32 maskR = shuffle_consts[index].right_mask;
+    index = 1 << ( 3 - index);
+    uint32 x = src & ~( maskL | maskR);
+    x |= ( ( src << index) & maskL) | ( ( src >> index) & maskR);
```

Comment:
Does this code exactly follow specification? If not, I would decompose it to "L" and "R" parts.

Reply:
36 and 37 strings exactly follow, 34 and 35 are additional because of the decreasing number of arguments in this function and making `uint32 bit_shuffle` less massive.